pkg/sdkserver/sdkserver.go

```diff
@@ -762,20 +784,196 @@ func (s *SDKServer) GetCounter(ctx context.Context, in *alpha.GetCounterRequest)
+			}
+			tmpReq.countSet = in.CounterUpdateRequest.CountSet
+			tmpReq.count = in.CounterUpdateRequest.Count
+			tmpReq.diffList = []int64{tmpReq.count}
+		case in.CounterUpdateRequest.CapacitySet: // Updated based on if client call is CapacitySet
+			tmpReq.capacitySet = in.CounterUpdateRequest.CapacitySet
+			tmpReq.capacity = in.CounterUpdateRequest.Capacity
+			// TODO: Do we need to check here that any previous changes to Count are still < Capacity?
+		default:
+			return nil, errors.Errorf("INVALID_ARGUMENT. Malformed CounterUpdateRequest: %v", in.CounterUpdateRequest)
+		}
+		s.gsCounterUpdates[name] = tmpReq
+	}
+
+	// Queue up the Update for later batch processing by updateCounters.
+	s.workerqueue.Enqueue(cache.ExplicitKey(string(updateCounters)))
+	return &alpha.Counter{}, nil
+}
+
+// updateCounter updates the Counters in the GameServer's Status with the batched update requests.
+func (s *SDKServer) updateCounter(ctx context.Context) error {
+	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
+		return errors.Errorf("%s not enabled", runtime.FeatureCountsAndLists)
+	}
+
+	gs, err := s.gameServer()
+	if err != nil {
+		return err
+	}
+	gsCopy := gs.DeepCopy()
+
+	s.logger.WithField("batchCounterUpdates", s.gsCounterUpdates).Debug("Batch updating Counter(s)")
+	s.gsUpdateMutex.Lock()
+	defer s.gsUpdateMutex.Unlock()
+
+	for name, ctrReq := range s.gsCounterUpdates {
+		counter, ok := gsCopy.Status.Counters[name]
+		if !ok {
+			continue
+		}
+		// If the Counter is the same as the original Counter stored in the counterUpdateRequest
+		// then no other changes have been made to this Counter, and we can overwrite the values.
+		if cmp.Equal(counter, ctrReq.counter) {
+			counter.Count = ctrReq.count
+			counter.Capacity = ctrReq.capacity
+		} else {
+			// If the Counter is NOT the same as the original Counter stored in the counterUpdateRequest
+			// then changes have been made to the Counter since we validated the incoming changes in
+			// UpdateCounter, and we need to verify if the batched changes can be fully applied, partially
+			// applied, or cannot be applied.
+			if ctrReq.capacitySet {
+				counter.Capacity = ctrReq.capacity
+			}
+			if ctrReq.countSet {
+				counter.Count, ctrReq.diffList = ctrReq.diffList[0], ctrReq.diffList[1:]
+			}
+			// If the diff can be applied, apply it. Otherwise loop through the requests.
+			newCnt := counter.Count + ctrReq.diff
+			if newCnt >= 0 && newCnt <= counter.Capacity {
+				counter.Count = newCnt
+			} else {
+				for _, req := range ctrReq.diffList {
+					newCnt = counter.Count + req
+					if newCnt >= 0 && newCnt <= counter.Capacity {
+						counter.Count = newCnt
+					}
+				}
+			}
+		}
+		gsCopy.Status.Counters[name] = counter
+		gs, err = s.gameServerGetter.GameServers(s.namespace).Update(ctx, gsCopy, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+		s.recorder.Event(gs, corev1.EventTypeNormal, "UpdateCounterBatchRequest",
```

Comment:
Nit:
```suggestion
		s.recorder.Event(gs, corev1.EventTypeNormal, "UpdateCounter",
```

Expect you were saving it for later,  but we could also do an additional Warning event if the count gets floored or ceiling-ed at 0 or `Capacity` .